### DISCUSSION TITLE
Explanation of performance & upatreaming

### DISCUSSION THREAD

**Discussion by adriangb** [pdm-project/pdm#788] [General]
From the benchmarks it looks like PDM performs much better than Poetry. Would it be possible to get a blog post or something summarizing where the performance improvement are coming from?

It would also be really cool if some of it could be ported over to Poetry (which is not going away anytime soon).

**Comment by frostming** [comment]
There is a post: https://frostming.com/2021/03-26/pm-review-2021/#performance
But I am not sure whether it covers your concern.

BTW may I know what benchmark result are you referring to?

**Reply by adriangb** [reply]
That's where I came from (those are the benchmark results I'm referring to). There is some info there on Pipenv (I guess it's cache design is flawed), but I do not see any information on _why_ PDM would be faster than Poetry (just the numbers)
Is it 1,000 micro optimizations? Does PDM take a drastically different approach to downloading large binaries just to inspect their sub-dependencies?

**Reply by frostming** [reply]
Here are some efforts that PDM takes to improve the performance, I am not clear of Poetry though.

1. Cache, HTTP requests, package dependencies, and wheel builds are all cached
2. Package hashes are extracted from the URL fragment without actually downloading them
3. Information in the lock file is used in dependency resolution to avoid unnecessary HTTP requests
4. The solver is different, which is the same as used by pip
5. Be aware that the performance testing in the post also includes installation. PDM uses an extremely simple installer and runs in parallel, which also contributes to the result.

**Reply by adriangb** [reply]
Very interesting. The use of info from URLs sounds like a smart move.

I now wonder how those numbers would look now that poetry has a parallelized resolver as well. So you by any chance have the source for your benchmarks? It might be interesting to put it in a `/benchmarks` directory for transparency and knowledge sharing.

**Reply by frostming** [reply]
From rough testing, it seems the latest Poetry(1.1.12) doesn't improve. I will paste the result and source code later #794

**Reply by frostming** [reply]
Here it is! https://pdm.fming.dev/dev/benchmark/

**Reply by adriangb** [reply]
Awesome, thank you for getting through that so quickly! It's interesting that there isn't much of a change for Poetry, I thought their [new installer](https://github.com/python-poetry/poetry/blob/master/CHANGELOG.md#110a3---2020-07-10) might make a difference.
